prometheus-exporters/netapp-exporter/charts/netapp-cap-exporter/aggregations/netapp-castellum.rules

```diff
@@ -54,3 +54,23 @@ groups:
 
       - record: manila_share_minimal_size_bytes_for_castellum
         expr: netapp_volume_provision_case_three * on (share_id) group_left max({__name__=~"netapp_volume_logical_used_bytes|netapp_volume_snapshot_used_bytes", volume_type="rw"}) by (share_id)
+
+      # If the `manila_share_exclusion_reasons_for_castellum` metric has entries, Castellum will ignore the respective share.
+      # The `reason` label in the final aggregation rule is used by Castellum for logging ignored shares.
+
+      # Scraping will fail on shares in state "offline" because their size is always reported as 0.
+      - record: netapp_volume_exclusion_reason_offline
+        expr: count(netapp_volume_total_bytes{project_id!="",share_id!="",volume_state="offline"}) by (project_id, share_id) > 0
+      - record: manila_share_exclusion_reasons_for_castellum
+        expr: label_replace(netapp_volume_exclusion_reason_offline, "reason", "volume_state = offline", "share_id", ".*")
+
+      # We want to ignore "shares" that are actually snapmirror targets. It's possible that we have both metrics with
+      # volume_type="dp" and other volume_type values, in which case Castellum will only use the non-dp metrics.
+      # This check is specifically about excluding shares that are *only* snapmirrors.
+      #
+      # NOTE: Not having any useful metrics at all is not a valid reason for ignoring the share.
+      # If we lack metrics about a share, we want to be alerted by the failing scrape.
+      - record: netapp_volume_exclusion_reason_dponly
+        expr: (count(netapp_volume_total_bytes{project_id!="",share_id!="",volume_type="dp"}) by (project_id, share_id) > 0) unless (count(netapp_volume_total_bytes{project_id!="",share_id!="",volume_type!="dp"}) by (project_id, share_id) > 0)
+      - record: manila_share_exclusion_reasons_for_castellum
+        expr: label_replace(netapp_volume_exclusion_reason_dponly, "reason", "volume_type = dp", "share_id", ".*")
```

Comment:
But do the dponly shares still show up in Manila? That's the main point here. This exclusion mechanism is used to skip shares that do exist in Manila, but do not have any usable metrics. Otherwise Castellum will throw errors when trying to collect size/usage metrics for these shares.

Reply:
dponly "shares" is not Manila idiom. What is confusing in the past is that there are one-to-many relationships in the prom time series, (a share may have several replicas). With the fix I mentioned earlier, only active replica is collected for a share, so they are one-to-one now. So Castellum can get the share id's and match them to `share_id` label in the metrics, and Castellum would not see "dp" volume at all.

Reply:
You seem to be thinking that Castellum is using NetApp metrics to find Manila shares. That is wrong; we are discovering shares via the Manila API. Let's use a concrete example: In eu-de-1, the `netapp_volume_exclusion_reason_dponly` metric from above has 54 matches, one of them being the share `486fe2e7-c323-4437-b64a-ad7c1552c477` (just picking a random example for illustration purposes). When I do `manila show 486fe2e7-c323-4437-b64a-ad7c1552c477`, I can see that this share also exists in the Manila API. Hence, Castellum would try to work with the share if autoscaling is enabled for the respective project, and then fail because the `for_castellum` metrics are not available for this share. This is why I have the exclusion rules in place.

I have amended the code comment to clarify this.

Reply:
@chuan137: such shares that are seemingly rw from manila perspective (having just one active replica) but indeed dp on netapp are part of a cross-region replication setup. E.g. the share example given by Stefan above got the replication source in another region: eu-de-2 (`61ecc7b3-ce0a-4bcf-ae3c-c70291a1dfa2` )

Reply:
thanks for explaining the case